src/library/scala/concurrent/BatchingExecutor.scala

```diff
@@ -55,56 +55,51 @@ trait Batchable {
   private[this] final val _tasksLocal = new ThreadLocal[Batch]()
 
-  private[this] final class Batch(capacity: Int) extends ArrayDeque[Runnable](capacity) with Runnable with BlockContext with (BlockContext => Unit) {
+  private[this] final class Batch(capacity: Int) extends ArrayDeque[Runnable](capacity) with Runnable with BlockContext with (BlockContext => Throwable) {
     private[this] final var parentBlockContext: BlockContext = _
 
     def this(r: Runnable, capacity: Int) = {
       this(capacity)
       addLast(r)
     }
 
-    final def executor: BatchingExecutor = BatchingExecutor.this
-
     // this method runs in the delegate ExecutionContext's thread
-    override final def run(): Unit = BlockContext.usingBlockContext(this)(this)
-
-    override final def apply(prevBlockContext: BlockContext): Unit = {
+    override final def run(): Unit = {
       //This invariant needs to hold: require(_tasksLocal.get eq null)
-      parentBlockContext = prevBlockContext
-      try {
-        _tasksLocal.set(this)
-        runAll()
-        _tasksLocal.remove() // Will be cleared in the throwing-case by runAll()
-      } finally {
-        parentBlockContext = null
-      }
+      _tasksLocal.set(this)
+      val failure = BlockContext.usingBlockContext(this)(this)
+      _tasksLocal.remove()
+      if (failure ne null)
+        throw handleRunFailure(failure)
     }
 
-    @tailrec private[this] final def runAll(): Unit = {
-      val next = pollLast()
-      if (next ne null) {
-        try next.run() catch {
-          case t: Throwable =>
-            parentBlockContext = null // Need to reset this before re-submitting it
-            _tasksLocal.remove() // If unbatchedExecute runs synchronously
-            handleRunFailure(t)
-         }
-        runAll()
+    override final def apply(prevBlockContext: BlockContext): Throwable = {
+      parentBlockContext = prevBlockContext
+      var failure: Throwable = null
+      try {
+        var r = pollLast()
+        while(r ne null) {
+          r.run()
+          r = pollLast()
+        }
+      } catch {
+        case t: Throwable => failure = t
       }
+      parentBlockContext = null
+      failure
     }
 
-    private[this] final def handleRunFailure(cause: Throwable): Nothing =
-      if (NonFatal(cause) || cause.isInstanceOf[InterruptedException]) {
-        try unbatchedExecute(this) catch {
+    private[this] final def handleRunFailure(cause: Throwable): Throwable =
+      if (!isEmpty && (NonFatal(cause) || cause.isInstanceOf[InterruptedException])) {
+        try { unbatchedExecute(this); cause } catch {
           case inner: Throwable =>
             if (NonFatal(inner)) {
               val e = new ExecutionException("Non-fatal error occurred and resubmission failed, see suppressed exception.", cause)
               e.addSuppressed(inner)
-              throw e
-            } else throw inner
+              e
+            } else inner
```

Comment:
should `inner` suppress `cause`?

Reply:
IMO yes—`inner` suppresses `cause` when `cause` is non-fatal and `inner` is fatal—this is to make sure that we propagate serious problems.